### PR TITLE
Upgrade kernels to mitigate Stack Clash

### LINKED ISSUE #3: use ubuntu specific kernel config
Since we are applying the ubuntu overlay we should also test and use their customized kernel config as the base for applying the grsec patch.
- [ ] Specify the airgap host OS
- [ ] Download and extract the current ubuntu specific kernel config
- [ ] Transfer the config to the airgap
- [ ] Verify the signature of the config file
- [ ] place the config file as `.config` in the directory where menuconfig is ran prior to patching the kernel with grsec.

### PR DESCRIPTION
**Pull request by conorsch**
Implements the mitigations described in https://github.com/freedomofpress/securedrop/issues/1861. These changes are SecureDrop-specific, and in addition to updating the grsecurity kernel patch version, also provide a sysctl modification to `vm.heap_stack_gap`, raising it 64KB -> 1MB, as [recommended by grsecurity](https://grsecurity.net/an_ancient_kernel_hole_is_not_closed.php). The sysctl options are rolled into the `securedrop-grsec` metapackage, same as the paxctl flags added via #91.

Also commits the latest kernel config used for SecureDrop, which is the default Ubuntu Trusty config with the grsecurity patch set applied with high security settings, as [indicated in the docs](https://github.com/freedomofpress/ansible-role-grsecurity/blob/a62731bf3c30ed6aab491e892693ce96ed501d87/docs/securedrop_metapackage.md), and thereby closes #3. We'll need to update that if and when we move to the 4.x kernel series.

### CODE REVIEW

File: roles/build-grsec-kernel/tasks/gpg_keys.yml [hunk 1]
```diff
@@ -25,3 +25,9 @@
   register: gpg_henriques_import
   changed_when: "'imported: 1' in gpg_henriques_import.stderr"
   when: grsecurity_build_include_ubuntu_overlay == true
+
+- name: Import Stefan Bader GPG key (Canonical/Ubuntu Kernel Team)
+  command: gpg --keyserver {{ grsecurity_build_gpg_keyserver }} --recv-keys E8675DEECBEECEA3
```
Comment (msheiny) [inline roles/build-grsec-kernel/tasks/gpg_keys.yml]: Thinking these keys should be a variable

Comment (psivesely) [reply roles/build-grsec-kernel/tasks/gpg_keys.yml]: Prefer the full fingerprint `DB5D7CCAF3994E3395DA4D3EE8675DEECBEECEA3`. While expensive, long key ID collisions are possible.

File: roles/build-grsec-metapackage/files/securedrop-sysctl-options.conf [hunk 2]
```diff
@@ -0,0 +1 @@
+vm.heap_stack_gap=1048576
```
Comment (msheiny) [inline roles/build-grsec-metapackage/files/securedrop-sysctl-options.conf]: I think should also be a default variable / template

**Review by psivesely — DISMISSED** [review]
The good:
* Correct permissions and ownerships.
* Correct sysctl parameters and use of `sysctl.d` file.
* Grsec configuration options listed in https://github.com/freedomofpress/securedrop/issues/1861#issuecomment-309636077 are enabled.
* Building headers too seems smart.
* Versioning seems in line with Debian packaging standards and upgrades should happen as expected.

The bad:
* Incorrect use of GPG.

**Comment by msheiny** [comment]
@conorsch - i really want to do a full test against this since its been a while since i ran this role -- im assuming this PR isnt urgent to get in since youve already built it but if my assumption is incorrect let me know

**Comment by conorsch** [comment]
@msheiny That's right, changes are already live and shipped. In order to test locally, I recommend something like:

```
vagrant up --no-provision grsec-build-securedrop
ansible-playbook -vv --diff -i .vagrant/provisioners/ansible/inventory/vagrant_ansible_inventory examples/build-grsecurity-kernel-securedrop.yml  --tags kernel
```

The manual playbook invocation is necessary for the `prompt` tasks to permit input from stdin, otherwise they'll be skipped and the role will fail.

**Comment by msheiny** [comment]
So two issues I ran into while deploying -- not really code specific but worth noting for the future:

* When you try to deploy the same kernel version from deb package directly -- you'll get a big scary error about existing modules in place. The only way I found to get past this was to either backup/move the existing dir or set `clobber_modules` via `/etc/kernel-img.conf`. I added commit that allows one to do that via the second option.
* When we deploy newer versions of kernels to over-write existing kernels with the same version... this is problematic because it makes it harder to check to see whats running during operations. Sure - I can check the dpkg version of whats installed but what if the system hasnt been rebooted yet and is still running an older version? I think we should always update kernel version that displays when you query from within the os. in other words,  `uname -r` output should always be incrementing when we update kernels internally.

File: roles/build-grsec-kernel/tasks/gpg_keys.yml (hunk 1 above)
Comment (msheiny) [reply roles/build-grsec-kernel/tasks/gpg_keys.yml]: Hey @conorsch - I think what @fowlslegs is requesting is a good call here (plus convert that to a default var)-- after that's done im 👍 up for margin' here

**Comment by conorsch** [comment]
@msheiny Revised GPG fingerprints to use longids, and moved the sysctl options to a vars-based template. Ready for re-review.

File: roles/build-grsec-kernel/tasks/gpg_keys.yml [hunk 3]
```diff
@@ -1,27 +1,13 @@
...
-
-- name: Import Luis Henriques GPG key (Canonical/LKM).
-  command: gpg --keyserver {{ grsecurity_build_gpg_keyserver }} --recv-keys DB74AEB8FDCE24FC
-  register: gpg_henriques_import
-  changed_when: "'imported: 1' in gpg_henriques_import.stderr"
+- name: Import GPG keys for building Linux kernel.
+  command: gpg --keyserver {{ grsecurity_build_gpg_keyserver }} --recv-keys "{{ item.fingerprint }}"
+  register: gpg_import_linux_pubkeys_result
+  changed_when: "'imported: 1' in gpg_import_linux_pubkeys_result.item.stderr"
+  with_items: "{{ grsecurity_build_gpg_keys }}"
```
Comment (msheiny) [inline roles/build-grsec-kernel/tasks/gpg_keys.yml:6]: `with_items`  💟 . I'm a sucker for a good ansible yaml loop

File: roles/build-grsec-kernel/defaults/main.yml [hunk 4]
```diff
@@ -70,3 +70,22 @@ grsecurity_build_fetch_packages_dest: "./"
 # The "test" patches do not require authentication or a subscription.
 grsecurity_build_download_username: ''
 grsecurity_build_download_password: ''
+
+# List of GPG keys required for building grsecurity-patched kernel.
+grsecurity_build_gpg_keys:
+  - name: Greg Kroah-Hartman GPG key (Linux stable release signing key)
+    fingerprint: 647F28654894E3BD457199BE38DBBDC86092693E
```
Comment (msheiny) [inline roles/build-grsec-kernel/defaults/main.yml:77]: Mmmmmmmmmmmmm fingerprinty

**Review by msheiny — APPROVED** [review]
i'll admit I didn't try to run this through this round - and only looked at the diff from last time. Thanks for addressing those fingerprint variable concerns 👍

**Pull request merged**